### PR TITLE
add LRFinder to CIFAR10 Example

### LINKED ISSUE #2004: Add LRFinder to CIFAR10 example
We already have an example for LRFinder: https://github.com/pytorch/ignite/blob/master/examples/notebooks/FastaiLRFinder_MNIST.ipynb , However we'd like to add LRFinder to another example. 
So we want to update this example: https://github.com/pytorch/ignite/tree/master/examples/contrib/cifar10 . 
The idea is to add an option as `with_lrfinder` and if true, setup and execute LRFinder and `apply_suggested_lr` method.

### PR DESCRIPTION
**Pull request by Ishan-Kumar2**
Fixes #2004
Description:
Added LRFinder to CIFAR10 Example.

(PS. apart from adding the LRFinder the rest of the changes are done by running black on the file.)

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by Ishan-Kumar2** [comment]
Also I noticed that the `num_warmup_epochs` is set by default to 4. 
So if you keep that to default and change the `num_epochs` to something < 4, the `PiecewizeLinear` scheduler [[here]](https://github.com/pytorch/ignite/blob/86cec1165441a945c93c938c23544ac52a17a170/examples/contrib/cifar10/main.py#L249)   throws an error as it should. The error message doesnt say the reason for this (`num_epochs` >= `num_warmup_epochs`).
Should I add a simple check?

**Comment by vfdev-5** [comment]
@Ishan-Kumar2 thanks for the PR, I let @KickItLikeShika to review it first.

> Also I noticed that the `num_warmup_epochs` is set by default to 4.
> So if you keep that to default and change the `num_epochs` to something < 4, the `PiecewizeLinear` scheduler [[here]](https://github.com/pytorch/ignite/blob/86cec1165441a945c93c938c23544ac52a17a170/examples/contrib/cifar10/main.py#L249) throws an error as it should. The error message doesnt say the reason for this (`num_epochs` >= `num_warmup_epochs`).
> Should I add a simple check?

Yes, you are right about that. We can add this check. On the other hand it does not make much sense to train on <4 epochs except of debuging the code :)

**Comment by Ishan-Kumar2** [comment]
haha you are right @vfdev-5 training for <4 epochs doesnt make sense, but lets help out the debuggers :)

File: examples/contrib/cifar10/main.py [hunk 1]
```diff
@@ -76,16 +76,21 @@ def training(local_rank, config):
...
     evaluator = create_evaluator(model, metrics=metrics, config=config)
     train_evaluator = create_evaluator(model, metrics=metrics, config=config)
 
+    # Finding optimum Learning Rate using FastaiLRFinder
+    if config["lr_finder"] is True:
+        lr_finder = FastaiLRFinder()
+        to_save = {"model": model, "optimizer": optimizer}
+        with lr_finder.attach(trainer, to_save, diverge_th=5.0) as trainer_with_lr_finder:
+            trainer_with_lr_finder.run(train_loader)
+            lr_finder.apply_suggested_lr(optimizer)
```
Comment (vfdev-5) [inline examples/contrib/cifar10/main.py:92]: This should be outside the context manager.

Comment (Ishan-Kumar2) [reply examples/contrib/cifar10/main.py]: fixed!

File: examples/contrib/cifar10/main.py [hunk 2]
```diff
@@ -76,16 +76,21 @@ def training(local_rank, config):
...
     evaluator = create_evaluator(model, metrics=metrics, config=config)
     train_evaluator = create_evaluator(model, metrics=metrics, config=config)
 
+    # Finding optimum Learning Rate using FastaiLRFinder
+    if config["lr_finder"] is True:
+        lr_finder = FastaiLRFinder()
+        to_save = {"model": model, "optimizer": optimizer}
+        with lr_finder.attach(trainer, to_save, diverge_th=5.0) as trainer_with_lr_finder:
+            trainer_with_lr_finder.run(train_loader)
+        lr_finder.apply_suggested_lr(optimizer)
```
Comment (KickItLikeShika) [inline examples/contrib/cifar10/main.py:92]: I think putting some logs here about finding the optimal learning rate would be nice

File: examples/contrib/cifar10/main.py [hunk 3]
```diff
@@ -76,16 +76,21 @@ def training(local_rank, config):
     trainer = create_trainer(model, optimizer, criterion, lr_scheduler, train_loader.sampler, config, logger)
```
Comment (KickItLikeShika) [inline examples/contrib/cifar10/main.py:76]: I hope you tried to run the example, because I think it won't work because of the `train_step` function, the return of `train_step` is a dict, and that's not supported for now with the LRFinder, the output must be of type float, and we are extending that in this PR by supporting tensors #2045 , so this must be edited https://github.com/pytorch/ignite/blob/ba8c63c358c96a5bdce428b708e7934c8e543a65/examples/contrib/cifar10/main.py#L329 
Check this method for more details https://github.com/pytorch/ignite/blob/master/ignite/handlers/lr_finder.py#L137

File: examples/contrib/cifar10/main.py [hunk 4]
```diff
@@ -76,16 +76,21 @@ def training(local_rank, config):
...
     # We define two evaluators as they wont have exactly similar roles:
     # - `evaluator` will save the best model based on validation score
     evaluator = create_evaluator(model, metrics=metrics, config=config)
     train_evaluator = create_evaluator(model, metrics=metrics, config=config)
 
+    # Finding optimum Learning Rate using FastaiLRFinder
+    if config["lr_finder"] is True:
+        lr_finder = FastaiLRFinder()
+        to_save = {"model": model, "optimizer": optimizer}
+        with lr_finder.attach(trainer, to_save, diverge_th=5.0) as trainer_with_lr_finder:
```
Comment (KickItLikeShika) [inline examples/contrib/cifar10/main.py:90]: `diverge_th` is 5.0 by default https://github.com/pytorch/ignite/blob/86cec1165441a945c93c938c23544ac52a17a170/ignite/handlers/lr_finder.py#L352

**Review by KickItLikeShika — COMMENTED** [review]
Thanks for the PR @Ishan-Kumar2, and sorry for not reviewing it quickly. Please try to run the example first, before committing the changes.

File: examples/contrib/cifar10/main.py (hunk 3 above)
Comment (vfdev-5) [reply examples/contrib/cifar10/main.py:76]: @KickItLikeShika I think we can use `output_transform` to transform dict to float, see https://github.com/pytorch/ignite/blob/05f2b558f6e10bd8b05b6157684c392e65473638/ignite/handlers/lr_finder.py#L370
We can keep the output as it is now and add `output_transform` to `lr_finder.attach`

File: examples/contrib/cifar10/main.py [hunk 5]
```diff
@@ -76,16 +76,21 @@ def training(local_rank, config):
...
-    }
+    metrics = {"Accuracy": Accuracy(), "Loss": Loss(criterion)}
 
     # We define two evaluators as they wont have exactly similar roles:
     # - `evaluator` will save the best model based on validation score
     evaluator = create_evaluator(model, metrics=metrics, config=config)
     train_evaluator = create_evaluator(model, metrics=metrics, config=config)
 
+    # Finding optimum Learning Rate using FastaiLRFinder
+    if config["lr_finder"] is True:
```
Comment (vfdev-5) [inline examples/contrib/cifar10/main.py:87]: Let's call the cmd argument as "use_lr_finder" instead of "lr_finder"

File: examples/contrib/cifar10/main.py (hunk 3 above)
Comment (KickItLikeShika) [reply examples/contrib/cifar10/main.py:76]: Nice idea, thanks @vfdev-5

Comment (Ishan-Kumar2) [reply examples/contrib/cifar10/main.py:76]: Great, thanks for the input. Should I add the `output_transform` in the `utils.py` or in `main` only?

File: examples/contrib/cifar10/main.py (hunk 4 above)
Comment (Ishan-Kumar2) [reply examples/contrib/cifar10/main.py]: fixed

File: examples/contrib/cifar10/main.py (hunk 5 above)
Comment (Ishan-Kumar2) [reply examples/contrib/cifar10/main.py]: Sure, I have made the change.

File: examples/contrib/cifar10/main.py (hunk 2 above)
Comment (Ishan-Kumar2) [reply examples/contrib/cifar10/main.py:94]: Yeah makes sense, I have added.

File: examples/contrib/cifar10/main.py (hunk 3 above)
Comment (Ishan-Kumar2) [reply examples/contrib/cifar10/main.py:76]: I have added it in `main.py` for now, let me know if it has to be moved.

File: examples/contrib/cifar10/main.py (hunk 2 above)
Comment (vfdev-5) [reply examples/contrib/cifar10/main.py:94]: @Ishan-Kumar2 I think we can just do : `output_transform=lambda output: output["batch loss"]` as we know what returns trainer. There is no need to do something generic.

Comment (Ishan-Kumar2) [reply examples/contrib/cifar10/main.py:94]: @vfdev-5 Okay, I have done that.

**Review by KickItLikeShika — APPROVED** [review]
Thanks for the updates @Ishan-Kumar2, LGTM!

File: examples/contrib/cifar10/main.py [hunk 6]
```diff
@@ -76,16 +76,24 @@ def training(local_rank, config):
...
-    }
+    metrics = {"Accuracy": Accuracy(), "Loss": Loss(criterion)}
 
     # We define two evaluators as they wont have exactly similar roles:
     # - `evaluator` will save the best model based on validation score
     evaluator = create_evaluator(model, metrics=metrics, config=config)
     train_evaluator = create_evaluator(model, metrics=metrics, config=config)
 
+    # Finding optimum Learning Rate using FastaiLRFinder
+    if config["use_lr_finder"] is True:
```
Comment (vfdev-5) [inline examples/contrib/cifar10/main.py:87]:
```suggestion
    if config["use_lr_finder"]:
```

**Review by vfdev-5 — APPROVED** [review]
Thanks for the PR @Ishan-Kumar2 ! Looks good

**Comment by Ishan-Kumar2** [comment]
Thanks a lot for the help and reviews @vfdev-5 and @KickItLikeShika. Excited to contribute more on ignite!

**Comment by KickItLikeShika** [comment]
Thanks for you contributions @Ishan-Kumar2, kindly check your Discord please!
